MaxText/layers/linears.py

```diff
@@ -376,22 +369,42 @@ def gmm(inputs, kernel, group_sizes):
 
       inputs = inputs.astype(self.dtype)
       kernel = kernel.astype(self.weight_dtype)
-
-      output = mblx.gmm(lhs=inputs,
-                        rhs=kernel,
+      output = mblx.gmm(lhs=inputs, 
+                        rhs=kernel, 
                         group_sizes=group_sizes,
-                        tiling=(512, 512, 512))
-
+                        preferred_element_type=jnp.bfloat16,
+                        tiling=tile_size)
+      
       if hs_shape[0] % pad_length:
         output = output[:hs_shape[0]]
       return output
 
-    layer_w0 = gmm(inputs, w0_kernel, group_sizes)
-    layer_w1 = gmm(inputs, w1_kernel, group_sizes)
-    layer_act = _convert_to_activation_function(mlp_activation)(layer_w0)
-    intermediate_layer = jnp.multiply(layer_act, layer_w1)
-    output = gmm(intermediate_layer, wo_kernel, group_sizes)
-    return output
+    @functools.partial(
+        shard_map.shard_map,
+        mesh=self.mesh,
+        in_specs=(
+              PartitionSpec(MESH_FSDP_AXIS, None, None),
+              PartitionSpec(MESH_FSDP_AXIS, None, None),
+              PartitionSpec(None, None, None),
+              PartitionSpec(None, None, None),
+              PartitionSpec(None, None, None),
+        ),
+        out_specs=PartitionSpec(MESH_FSDP_AXIS, None, None),
```

Comment:
This should also be all batch dimensions (DP, FSDP, FSDP_TRANSPOSE), nothing special about the FSDP one I believe - you should be able to use the same nn.logical_to_physical API

Reply:
Updated. Thanks!